docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,325 @@
+.. _config_http_filters_lua:
+
+Lua
+===
+
+.. attention::
+
+  The Lua scripting HTTP filter is **experimental**. Use in production at your own risk. It is
+  being released for initial feedback on the exposed API and for further development, testing,
+  and verification. This warning will be removed when we feel that the filter has received enough
+  testing and API stability to call it generally production ready.
+
+Overview
+--------
+
+The HTTP Lua filter allows `Lua <https://www.lua.org/>`_ scripts to be run during both the request
+and response flows. `LuaJIT <http://luajit.org/>`_ is used as the runtime. Because of this, the
+supported Lua version is mostly 5.1 with some 5.2 features. See the `LuaJIT documentation
+<http://luajit.org/extensions.html>`_ for more details.
+
+The filter only supports loading Lua code in-line in the configuration. If local filesystem code
+is desired, a trivial in-line script can be used to load the rest of the code from the local
+environment.
+
+The design of the filter and Lua support at a high level is as follows:
+
+* All Lua environments are :ref:`per worker thread <arch_overview_threading>`. This means that
```

Comment:
Do we want persistent state in Lua that extends beyond a single request? Why not have C++ objects hold the per-worker persistent state, accessible via a FFI API, and have Lua state ephemeral?

Reply:
I think it would be better to avoid FFI for the public API.  It doesn't make for nice Lua APIs, because you can't pass a table into an FFI function and do anything meaningful with it.  Or at least that's my recollection of the luajit ffi; it's been a few years since I've used it.

But I do like the idea of being able to store state across requests

Reply:
There are a few things at play here:

- FFI vs. not. Based on my research I recommend we do not use FFI at all. It's not safe and programmers can cause crashes. I would rather we sacrifice perf for a full jail that is safe. (FFI also I believe can really mess up the jitter and in some cases can produce slower code).

- I did quite a lot of research in terms of how to hide state between requests. First, Lua is not thread safe, so basically the high level state must be per thread. Creating a new state for each request would have a couple of downsides: 1) We would need to re-JIT for every request. 2) There would be no support at all for global/init data. I think it's important for us to allow people to have global state. In the future, I think we could support true global state of some type (maybe just string map, etc.).

Reply:
Also, the whole DeathRef situation is to make sure that people cannot incorrectly use objects across requests, but they can store global/expensive state if they really need to.

Reply:
OK, thanks for the followup. I'm new to Lua embeddings, just wanted to make sure the design space is fully covered, which knowing the folks involved, was a given anyway ;)